[AIRFLOW-1517] kubenetes operator

dimberman opened the pull request:
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-XXX


### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

The kubernetes pod operator will allow users to launch arbitrary pods on kubernetes. They can load an image, environment variables, volume mount, and necessary command to launch.

This PR also lays a lot of the necessary scaffolding for the soon-to-be released kubernetes executor (things like the ability to launch pods/secrets, etc.)


### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:


### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

dimberman committed:
[AIRFLOW-1517] Kubernetes Operator

foxish commented on scripts/ci/requirements.txt:
```diff
@@ -1,5 +1,5 @@
 #
-# Licensed under the Apache License, Version 2.0 (the "License");
+# Licensed RUN_KUBE_INTEGRATIONunder the Apache License, Version 2.0 (the "License");
```
Looks like an unintended change here

foxish commented on airflow/contrib/kubernetes/secret.py:
```diff
@@ -0,0 +1,36 @@
+
+
+class Secret:
+    """Defines Kubernetes Secret Containers"""
```
Is this referring to a Secret Volume Mount?

dimberman replied:
@foxish Yes, this allows users to mount existing secrets to any pod they are deploying.

foxish replied:
I think the comment/doc string should be changed in that case - to "Defines Kubernetes Secret Volumes" since it's not a container by itself.

foxish reviewed (commented):
Seems like a good start. Couple of minor comments from an initial pass.

grantnicholas commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -0,0 +1,64 @@
+            client = kube_client.get_kube_client(in_cluster=False)
+            gen = pod_generator.PodGenerator()
+
+            pod = gen.make_pod(namespace=self.namespace,
```
The `make_pod` function provides a lot of configurable options but we do not expose them as part of the operator. I think it makes sense to expose these options. (The big one I'm thinking of is resource limits but others matter too)

grantnicholas commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -0,0 +1,64 @@
+                               )
+
+            launcher = pod_launcher.PodLauncher(client)
+            final_state = launcher.run_pod(pod)
```
synchronous `run_pod` should probably take a configurable timeout to start (and have this exposed as part of the operator). if a pod has not started up in ${N} seconds the pod is probably never going to start up.

grantnicholas commented on tests/contrib/operators/test_kubernetes_pod_operator.py:
```diff
@@ -0,0 +1,34 @@
+from airflow import AirflowException
+
+
+class KubernetesPodOperatorTest(unittest.TestCase):
```
Does it make sense to add the kubernetes integration testing setup so we can actually test this operator? We can always add it in a later PR if that bloats this PR too much.

dimberman replied:
I think it makes sense. I think these smaller PRs are a good opportunity to up our unit/integration testing.

grantnicholas commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -0,0 +1,64 @@
+        raise AirflowException("trying to except")
+        try:
+
+            client = kube_client.get_kube_client(in_cluster=False)
```
Probably want to let people run this operator from inside/outside the cluster as well.

dimberman replied:
Yeah, this goes under the WIP section. good to comment here to make sure I fix this

grantnicholas commented on airflow/contrib/kubernetes/pod_generator.py:
```diff
@@ -0,0 +1,290 @@
+        self.init_containers = []
+        self.secrets = []
+
+    def set_environment(self, env):
```
Personal gripe time (feel free to ignore):
getters/setters in python that just wrap a property are usually bad practice. is there a good reason we need these (I don't see where this setter is used).

grantnicholas commented on airflow/contrib/kubernetes/pod_launcher.py:
```diff
@@ -0,0 +1,103 @@
+            time.sleep(2)
+        return self.read_pod(pod)
+
+        # for event in self._watch.stream(self.read_pod(pod), pod.namespace):
```
Reminder to delete this. I think polling seems like a good solution for the operator (not the scheduler).

benjigoldberg committed:
[AIRFLOW-1517] Remove authorship of secrets and init container

benjigoldberg committed:
[AIRFLOW-1517] Restore authorship of secrets and init container

Commit:
[AIRFLOW-1517] Add minikube for kubernetes integration tests

Add better support for minikube integration tests; By default minikube integration tests will run with kubernetes 1.7 and kubernetes 1.8

Commit:
[AIRFLOW-1517] Add minikube for kubernetes integration tests

Add better support for minikube integration tests; By default minikube integration tests will run with kubernetes 1.7 and kubernetes 1.8

Commit:
[AIRFLOW-1517] Remove authorship of resources

Collaboration authors got destroyed when splitting up a PR, this commit removes code which will be readded in the next commit to restore authorship

Commit:
[AIRFLOW-1517] Remove authorship of resources

Collaboration authors got destroyed when splitting up a PR, this commit removes code which will be readded in the next commit to restore authorship

Commit:
[AIRFLOW-1517] Restore authorship of resources

Collaboration authors got destroyed when splitting up a PR, this commit adds back in the code which was be removed in the previous commit to restore authorship

Commit:
[AIRFLOW-1517] Restore authorship of resources

Collaboration authors got destroyed when splitting up a PR, this commit adds back in the code which was be removed in the previous commit to restore authorship

dimberman committed:
Revert "[AIRFLOW-1517] Add minikube for kubernetes integration tests"

This reverts commit 0197931609685a98181387014f7c8f3b5cd5f9a8.

bolkedebruin commented on tests/contrib/minikube_tests/test_kubernetes_pod_operator.py:
```diff
@@ -0,0 +1,78 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Please use this license blob everywhere else as well. The other one is outdated and not accepted by Apache (anymore).

bolkedebruin commented on airflow/contrib/kubernetes/kube_client.py:
```diff
@@ -0,0 +1,32 @@
+            config.load_kube_config()
+            return client.CoreV1Api()
+        except NotImplementedError:
+            NotImplementedError(
```
Please don't swallow the exception. And from the looks of it you want to reraise it as a AirflowConfigException?

dimberman committed:
[AIRFLOW-1517] Created more accurate failures for kube cluster issues

dimberman committed:
[AIRFLOW-1517] fixed license issues

bolkedebruin commented on airflow/contrib/kubernetes/kube_client.py:
```diff
@@ -0,0 +1,30 @@
+    else:
+        config.load_kube_config()
+        return client.CoreV1Api()
+
```
Two blank lines required

bolkedebruin commented on scripts/ci/kubernetes/minikube/start_minikube.sh:
```diff
@@ -0,0 +1,81 @@
+#  Licensed to the Apache Software Foundation (ASF) under one   *
```
Why the asterisk at eol?

bolkedebruin commented on scripts/ci/kubernetes/minikube/start_minikube.sh:
```diff
@@ -0,0 +1,81 @@
+#  specific language governing permissions and limitations      *
+#  under the License.                                           *
+
+#!/usr/bin/env bash
```
At the top plz

bolkedebruin commented on scripts/ci/kubernetes/minikube/start_minikube.sh:
```diff
@@ -0,0 +1,81 @@
+fi
+#
+
+curl -Lo minikube https://storage.googleapis.com/minikube/releases/v0.24.1/minikube-linux-amd64 && chmod +x minikube
```
You probably want to store this in travis’ cache. See the Hadoop scripts as an example

dimberman replied:
@grantnicholas could you please set this part up?

grantnicholas replied:
@dimberman sure will do.

dimberman committed:
[AIRFLOW-1517] started documentation of k8s operator

dimberman committed:
[AIRFLOW-1517] addressed PR comments

bolkedebruin commented on setup.py:
```diff
@@ -162,6 +162,8 @@ def write_version(filename=os.path.join(*['airflow',
 cloudant = ['cloudant>=0.5.9,<2.0'] # major update coming soon, clamp to 0.x
 redis = ['redis>=2.10.5']
+kubernetes = ['kubernetes>=3.0.0',
+              'cryptography>=2.0.0']
```
Probably you want to update the crypto dependency (see higher up) and reference it from here. I know it could affect non-k8s installs but I don’t think it is smart to maintain two dependencies.

cmkimerer commented on docs/kubernetes.rst:
```diff
@@ -0,0 +1,36 @@
++--------------+----------------------------------------------------------------+---------------+
+| ``@image``    | docker image you wish to launch. Defaults to dockerhub.io, but fully qualified URLS will point to custom repositories |
++--------------+----------------------------------------------------------------+---------------+
+| ``@cmnds``  | To start a task in a docker image, we need to tell it what to do. the cmds array is the space seperated bash command that will define the task completed by the container           |
```
cmnds -> cmds

cmkimerer commented on scripts/ci/kubernetes/minikube/start_minikube.sh:
```diff
@@ -0,0 +1,81 @@
+# under the License.
+
+# Guard against a kubernetes cluster already being up
+kubectl get pods &> /dev/null
```
The assumption here is that kubectl isn't configured to talk to another cluster and that the other cluster isn't the default.  I don't think there is a better way to prove you're talking to minikube as intended, i guess you could look at the current context?

There's not a great way to parse that the config context stuff since you can't (or previously couldn't) change output types to something easily parsable.
  

Some other options here would be to ensure minikube is running using something like `minikube status` and that kubectl is configured to talk to the minikube cluster with `minikube update-context` ?

ethernetdan replied:
To prevent collisions, `minikube profile` could be used to set a kubectl context name that's unique enough to be unlikely accidentally used. Then `kubectl config set-context` (or alias kubectl) could be used to ensure that minikube is used.

dimberman replied:
@grantnicholas I think @cmkimerer and @ethernetdan have a good point here, do you want to set this up? (I can do it this weekend if you're busy)

foxish replied:
Drive-by thought - Isn't this meant to run under travis only? In that case, there is complete control over the environment and we can be sure there isn't a different k8s cluster configured also. 

For the cleanup, we usually run the tests in a separate randomly generated namespace every time and clean up the namespace which ensures everything - i.e. secrets, services, pods, etc get cleaned up.

dimberman replied:
fair point. The other consideration is making it as easy as possible for users to unit test if they want to add features, but even then there's no reason they HAVE to use minikube (at least none of our unit tests are minikube specific).

cmkimerer replied:
Given the comment here: https://github.com/apache/incubator-airflow/pull/2853/files#diff-a6fe26ef61545b07f21f31c42a69d4e3R22, it looks like the desired way to do development work on this also calls this script.

However, if you had just started minikube though as that linked documentation also suggests, you also have likely just re-configured your kubectl's context.

It does seem like the `setup_kubernetes.sh` script isn't all that helpful in setting up the development environment as it is now though.  Maybe it was intended to be more in-depth, or was at one point and lost (unneeded) functionality as this patch evolved?

foxish replied:
Yeah, if the goal is also to support user-run testing, we could separate out the actual testing from the environment setup into a script which takes `master` as a parameter. Not sure if that's worth the effort right now.

dimberman replied:
Eh, I'm fine to put that on the roadmap, but I really want to get people trying this out as an MVP/alpha

grantnicholas replied:
@cmkimerer While the script is technically only supposed to be used for travis, it's nice to have a way to setup a "standard" development environment. I know when we were all collaborating it was hard to prevent "works on my machine" issues with different kubernetes setups so having one standard setup enforced a certain amount of reliability. 

You are right that much of the functionality of the script was removed as it only applies to the kubernetes executor and not the operator. So the script *is* useful but you can't really see that right now until the executor PR is opened. 

I think we probably want to make this a bit more robust, but probably not a blocker for a first release (imo).

cmkimerer replied:
@grantnicholas Yea, I'm definitely not trying to hold up this from landing (as I'm waiting on this and the executor PR 😄 to land), just figured I'd call out the fact this is potentially dangerous and is something we should smooth out [in a future release]

cmkimerer commented on tests/contrib/minikube_tests/test_kubernetes_pod_operator.py:
```diff
@@ -0,0 +1,78 @@
+        k.execute(None)
+
+    def test_faulty_image(self):
+        bad_image_name = "foobar"
```
I guess the assumption here is that foobar will never be a valid image name in docker.io repository? Probably a safe assumption, but i'll call it out anyway 😄

ethernetdan replied:
The image could also be in a domain that should never allow discovery of an image, ie. `example.com/bad/image`

dimberman replied:
@ethernetdan should we limit what websites people can pull images from?

@cmkimerer fair point :), I can make it a bit more obscure to ensure it's never a hit (worst case it's a quick fix to change the image name for this test)

cmkimerer commented on airflow/contrib/kubernetes/pod.py:
```diff
@@ -0,0 +1,94 @@
+    :param result: The result that will be returned to the operator after
+                   successful execution of the pod
+    :type result: any
+    """
```
Looks like your docstring got out of sync with the kwargs, not sure if that's intentional?

cmkimerer commented on airflow/contrib/kubernetes/pod.py:
```diff
@@ -0,0 +1,94 @@
+                   successful execution of the pod
+    :type result: any
+    """
+    pod_timeout = 3600
```
This appears to be unused?

dimberman replied:
Good call, will remove

kevincvlam commented on airflow/contrib/kubernetes/pod_launcher.py:
```diff
@@ -0,0 +1,121 @@
+        # type: (Pod) -> State
+
+        while self.pod_is_running(pod):
+            self.log.info("Pod {} has state {}".format(pod.name, State.RUNNING))
```
Would it make sense to pass logs from the pod along to airflow, so it appears in the task logs? Eg. so you can track the progress of the worker pod in the airflow web UI.

dimberman replied:
@kevincvlam good point. Should be pretty easy to use the kubernetes client to live-log the results of the pod. This would also handle the logging issues for cases where people don't have ELK (though we probably want this to be optional in case people have their own logging solutions)

grantnicholas replied:
@kevincvlam That's a good question and there are a bunch of potential solutions. (You can use a read/write many persistent volume and mount the logs folder into each worker, you can intercept the logs from the kubernetes api itself, you can use a custom airflow logging solution). 

All of them have tradeoffs, I think we chose not to do anything because there were so many options and not a clear winner. I'm open to exploring this more as I think it's especially important for the executor.

ethernetdan commented on airflow/contrib/kubernetes/pod_generator.py:
```diff
@@ -0,0 +1,281 @@
+                           ):
+        """
+
+        Adds an init container to the launched pod. useful for pre-
```
cliffhanger 😄

kevincvlam commented on airflow/contrib/kubernetes/kubernetes_request_factory/kubernetes_request_factory.py:
```diff
@@ -0,0 +1,165 @@
+
+        req['spec']['containers'][0]['resources'] = {}
+
+        if pod.resources.has_requests():
```
Is it easy to support more arbitrary pod resources requests/limits? For example, 'nvidia.com/gpu'. Would be nice to be able to request GPU resources.

dimberman replied:
@kevincvlam what would you recommend for the user interface for that? I'd want to avoid forcing users to create a dict.

grantnicholas replied:
@kevincvlam 
There is definitely a tradeoff here between extensibility and type-safety/readability of the api. 

The simplest api would be to let users write their pod as a nested dict that gets directly translated into kubernetes json, but that can be painful for some tasks. The api we provided is more pythonic, with the downside that exposing new features of kubernetes requires adding more code in the "translation layer". 

I'm open to both approaches, I think there can be a case made for both.

kevincvlam replied:
@dimberman @grantnicholas makes sense, and good points. I think @grantnicholas highlighted the trade-off well, and I'm not sure what approach is best. Might mull it over a bit more.

In the case that it's decided to avoid forcing users to create a dict, I'd request that the 'nvidia.com/gpu' resource type be added in the "translation layer" :-).

dimberman replied:
@kevincvlam Would you be willing to either PR this in once this PR is merged or leave a quick issue on https://github.com/bloomberg/airflow/issues? I think this is a good feature but It'll be a lot easier to add things in once the initial PR is merged.

kevincvlam replied:
@dimberman Possibly.. for now created issue: https://github.com/bloomberg/airflow/issues/42. +1, this makes sense as a follow-up PR. Thanks!

ethernetdan commented on airflow/contrib/kubernetes/pod_launcher.py:
```diff
@@ -0,0 +1,121 @@
+        state = self._task_status(self.read_pod(pod))
+        return state != State.SUCCESS and state != State.FAILED
+
+    def read_pod(self, pod):
```
May want to raise a more helpful exception if this 404s

ethernetdan commented on scripts/ci/kubernetes/minikube/start_minikube.sh:
```diff
@@ -0,0 +1,81 @@
+  do
+    echo "------- Running kubectl get pods -------"
+    STDERR=$(kubectl get pods  2>&1 >/dev/null)
+    if [ $? -ne 1 ]; then
```
May want to check for `0` here to allow exit codes other than `1`

Commit:
[AIRFLOW-1517] Kubernetes operator PR fixes

Fix python flake8 linting issues and AIRFLOW license issues
